core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -104,10 +110,13 @@
   // TODO(gpang): remove this set, and just use the indexes.
   private final ConcurrentHashSet<T> mObjects = new ConcurrentHashSet<>(8, 0.95f, 8);
   /**
-   * Map from {@link FieldIndex} to the index. An index is a map from index value to set of
-   * objects with that index value.
+   * Map from {@link FieldIndex} to the index. An index is a map from index value to one or a set of
+   * objects with that index value. A unique index is an index where each index value only maps to
+   * one object. A non-unique index is an index where an index value can map to one or more objects.
    */
-  private final Map<FieldIndex<T>, ConcurrentHashMap<Object, ConcurrentHashSet<T>>> mIndexMap;
+  private final Map<UniqueFieldIndex<T>, ConcurrentHashMap<Object, T>> mIndexMapUnique;
```

Comment:
I think we could simplify this class by having field indices manage their own indexed values. What we currently call a field index is actually more of a field index function. An index class could have add, get, and remove methods, and Unique/Nonunique field indices would implement these slightly differently so that unique index returns either an empty set or a singleton set. An index would take an index field function in its constructor and this class would hold a list of indices.